vault/identity_store_oidc.go

```diff
@@ -0,0 +1,1344 @@
+		key.Algorithm = d.Get("algorithm").(string)
+	}
+
+	if key.Algorithm != "RS256" {
+		return logical.ErrorResponse("unknown signing algorithm %q", key.Algorithm), nil
+	}
+
+	// Update next rotation time if it is unset or now earlier than previously set.
+	nextRotation := time.Now().Add(key.RotationPeriod)
+	if key.NextRotation.IsZero() || nextRotation.Before(key.NextRotation) {
+		key.NextRotation = nextRotation
+	}
+
+	// generate keys if creating a new key
+	if key.SigningKey == nil {
+		signingKey, err := generateKeys(key.Algorithm)
+		if err != nil {
+			return nil, err
+		}
+
+		key.SigningKey = signingKey
+		key.KeyRing = append(key.KeyRing, &expireableKey{KeyID: signingKey.Public().KeyID})
+
+		if err := saveOIDCPublicKey(ctx, req.Storage, signingKey.Public()); err != nil {
+			return nil, err
+		}
+	}
+
+	// store named key
+	entry, err = logical.StorageEntryJSON(namedKeyConfigPath+name, key)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := req.Storage.Put(ctx, entry); err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+// handleOIDCReadKey is used to read an existing key
+func (i *IdentityStore) pathOIDCReadKey(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	name := d.Get("name").(string)
+
+	entry, err := req.Storage.Get(ctx, namedKeyConfigPath+name)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return logical.ErrorResponse("no named key found at %q", name), nil
+	}
+
+	var storedNamedKey namedKey
+	if err := entry.DecodeJSON(&storedNamedKey); err != nil {
+		return nil, err
+	}
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"rotation_period":  int64(storedNamedKey.RotationPeriod.Seconds()),
+			"verification_ttl": int64(storedNamedKey.VerificationTTL.Seconds()),
+			"algorithm":        storedNamedKey.Algorithm,
+		},
+	}, nil
+}
+
+// handleOIDCDeleteKey is used to delete a key
+func (i *IdentityStore) pathOIDCDeleteKey(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	targetKeyName := d.Get("name").(string)
+
+	// it is an error to delete a key that is actively referenced by a role
+	roleNames, err := req.Storage.List(ctx, roleConfigPath)
+	if err != nil {
+		return nil, err
+	}
+
+	var role *role
+	rolesReferencingTargetKeyName := make([]string, 0)
+	for _, roleName := range roleNames {
+		entry, err := req.Storage.Get(ctx, roleConfigPath+roleName)
+		if err != nil {
+			return nil, err
+		}
+		if entry != nil {
+			if err := entry.DecodeJSON(&role); err != nil {
+				return nil, err
+			}
+			if role.Key == targetKeyName {
+				rolesReferencingTargetKeyName = append(rolesReferencingTargetKeyName, role.Name)
+			}
+		}
+	}
+
+	if len(rolesReferencingTargetKeyName) > 0 {
+		errorMessage := fmt.Sprintf("unable to delete key %q because it is currently referenced by these roles: %s",
+			targetKeyName, strings.Join(rolesReferencingTargetKeyName, ", "))
```

Comment:
Is this information that otherwise would be behind a different ACL?

Reply:
Yeah. We went back and forth on that, eventually thinking that benefit of the feedback (vs. trying to figure out the offending roles), esp given that the person is already authorized to set things up, outweighed the printing of the roles. We can remove this detail if you want.

Reply:
Discussed with @jefferai and we'll leave as is right now.